proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+
+.. [#] See `What should happen if multiple modules are defined
+       with the same name?`_ for discussion.
+
+::
+
+   import A as M
+   import B as M
+
+means the same as
+
+::
+
+   module M (module A(..), module B(..)) where
+       import A
+       import B
+
+This retains the current behaviour with respect to clashing names, since
+Haskell does not issue ambiguity warnings on import, only on usage. Thus
+the hypothetical definition of ``M`` is error-free even if ``A`` and
+``B`` export ``n``, but an attempt to use ``M.n`` will result in an
+ambiguity error. [#]_
+
+.. [#] This explication of the semantics uses a local module definition,
+   but we could wire the behaviour in as a special case if we decide
+   not to support local modules.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous section allows us to refer to modules by name, but not to
+export those names.
+
+We propose adding a language extension ``ExportModuleNames`` which adds the following
+export specifiers:
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and some of its names.
+- ``module M(..)`` exports ``M`` and all of its names. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+This ``ExportModuleNames`` behaviour is the “right” behaviour as far as
+this proposal is concerned, but we gate it behind a language extension
+for backwards compatibility.
+
+Local modules
+~~~~~~~~~~~~~
+
+We propose adding a language extension ``LocalModules``. This allows
+module definitions within modules, with the same syntax as in [ghc283]_. Local
+modules are very like file modules, in particular:
+
+- They can have imports.
+
+  - The imports can only target modules which can be referred to
+    with a qualified name in the current scope. Their target module name is *never*
+    interpreted as a dotted name.
+
+- They can have nested module definitions.
+- They can have export specifiers.
+- They can *not* have dotted names.
+
+The rules for local modules are mostly the same as for file modules,
+but:
+
+- A local module definition adds a binding for its declared name to
+  its enclosing scope, as usual for definitions.
+- Bindings from the
+  enclosing scope are in scope in the local module, but are shadowed if
+  there is a local definition of the same name. [#]_
+
+.. [#] See the `Recursion`_ section for some subtleties.
+
+Local modules *cannot* import file modules. This ensures that we can
+always compute the dependency graph of the compilation units of a
+Haskell program by looking at the top-level imports alone. [#]_
+
+.. [#] See `Do we need different syntax for "local" imports?`_ for discussions
+   of the consequences.
+
+Module aliases
+~~~~~~~~~~~~~~
+
+We propose adding a language extension ``ModuleAliases``. This allows a
+new kind of top-level definition: ``module <name> = <qualname>``
```

Comment:
What is your proposal on the restriction on the dotted-ness of the new binding?

Reply:
I say it shouldn't be allowed. if it was, it should be sugar for nested local modules as Nix does with ` { a.b.c.d = ...; }`

Reply:
Yes, no dots.